controllers/keda/metrics_adapter_controller.go

```diff
@@ -43,3 +91,47 @@ func (r *MetricsScaledObjectReconciler) SetupWithManager(mgr ctrl.Manager, optio
 		WithOptions(options).
 		Complete(r)
 }
+
+func (r *MetricsScaledObjectReconciler) addToMetricsCache(namespacedName string, metrics []string) {
+	scaledObjectsMetricsLock.Lock()
+	defer scaledObjectsMetricsLock.Unlock()
+	scaledObjectsMetrics[namespacedName] = metrics
+	extMetrics := populateExternalMetrics(scaledObjectsMetrics)
+
+	r.ExternalMetricsInfoLock.Lock()
+	(*r.ExternalMetricsInfo) = extMetrics
+	r.ExternalMetricsInfoLock.Unlock()
+}
+
+func (r *MetricsScaledObjectReconciler) removeFromCache(namespacedName string) {
+	scaledObjectsMetricsLock.Lock()
+	defer scaledObjectsMetricsLock.Unlock()
+	delete(scaledObjectsMetrics, namespacedName)
+	extMetrics := populateExternalMetrics(scaledObjectsMetrics)
+
+	// the metric could have been already removed by the previous call
+	// in this case we don't have to rewrite r.ExternalMetricsInfo
+	changed := false
+	r.ExternalMetricsInfoLock.RLock()
+	if len(*r.ExternalMetricsInfo) != len(extMetrics) {
+		changed = true
+	}
+	r.ExternalMetricsInfoLock.RUnlock()
+
+	if changed {
+		r.ExternalMetricsInfoLock.Lock()
+		(*r.ExternalMetricsInfo) = extMetrics
+		r.ExternalMetricsInfoLock.Unlock()
+	}
+}
+
+func populateExternalMetrics(scaledObjectsMetrics map[string][]string) []provider.ExternalMetricInfo {
+	externalMetrics := []provider.ExternalMetricInfo{}
```

Comment:
You sure you want a slice of structs rather than pointers? That's a lot more memcpy going on when this gets passed around.

Reply:
Slice of pointers to the struct? I return it this way because that's what the interface wants to return and didn't wanted to do another conversion then again. But if you think that's better I could try to change it https://github.com/kedacore/keda/blob/12452b67dddf41364762ebe2146e58595b4ef5b6/pkg/provider/provider.go#L147

Reply:
It's probably fine, ignore me. The slice itself will be fast to move around and that's all that matters :)

Reply:
Or I can return pointer to slice if that could be any better.
I am glad for these kind of reviews :)

Reply:
Nah, a reference to a slice (or map) is already pretty tiny, it's a pointer to the memory block and two ints for the len and cap. This would be a deeper design change for the interface for it to matter, when you `range` over any `[]struct` it's going to copy the whole struct each time through the loop. Technically with a `[]*struct` it's copying too but at least the pointer usually smaller than the struct it's pointing to.

Reply:
Yeah, i will keep it as it is atm then. Thanks for the review!